var/spack/repos/builtin/packages/rust/package.py

```diff
@@ -4,59 +4,426 @@
-          '--sysconfdir=%s' % join_path(prefix, 'etc/')
-          ]
-
-        configure(*configure_args)
-
-        # Build system defaults to searching in the same path as Spack's
-        # compiler wrappers which causes the build to fail
-        filter_file(
-            '#ar = "ar"',
-            'ar = "%s"' % join_path(spec['binutils'].prefix.bin, 'ar'),
-            'config.toml')
+        target = self.get_rust_target()
+        # See the NOTE above the resource loop - should be host architecture,
+        # not target aarchitecture if we're to support cross-compiling.
+        bootstrapping_install = Executable(
+            './spack_bootstrap_stage/rust-{version}-{target}/install.sh'
+            .format(
+                version=spec.version,
+                target=target
+            )
+        )
+        # install into the staging area
+        bootstrapping_install('--prefix={0}'.format(
+            join_path(self.stage.source_path, 'spack_bootstrap')
+        ))
+
+        boot_bin = join_path(self.stage.source_path, 'spack_bootstrap/bin')
+
+        # Always build rustc and cargo
+        tools = ['rustc', 'cargo']
+        # Only make additional components available in 'rust-bootstrap'
+        if '+rustfmt' in self.spec:
+            tools.append('rustfmt')
+        if '+analysis' in self.spec:
+            tools.append('analysis')
+        if '@1.33: +clippy' in self.spec:
+            tools.append('clippy')
+        if '+rls' in self.spec:
+            tools.append('rls')
+        if '+src' in self.spec:
+            tools.append('src')
+
+        ar = which('ar', required=True)
+
+        # build.tools was introduced in Rust 1.25
+        tools_spec = 'tools={0}'.format(tools) if '@1.25:' in self.spec else ''
+        # This is a temporary fix due to rust 1.42 breaking self bootstrapping
+        # See: https://github.com/rust-lang/rust/issues/69953
+        #
+        # In general, this should be safe because bootstrapping typically
+        # ensures everything but the bootstrapping script is warning free for
+        # the latest set of warning.
+        deny_warnings_spec = \
+            'deny-warnings = false' if '@1.42.0' in self.spec else ''
+
+        with open('config.toml', 'w') as out_file:
+            out_file.write("""\
+[build]
+cargo = "{cargo}"
+rustc = "{rustc}"
+docs = false
+vendor = true
+extended = true
+verbose = 2
+{tools_spec}
+
+[rust]
+channel = "stable"
+rpath = true
+{deny_warnings_spec}
+
+[target.{target}]
+ar = "{ar}"
+
+[install]
+prefix = "{prefix}"
+sysconfdir = "etc"
+""".format(
+                cargo=join_path(boot_bin, 'cargo'),
+                rustc=join_path(boot_bin, 'rustc'),
+                prefix=prefix,
+                target=target,
+                deny_warnings_spec=deny_warnings_spec,
+                ar=ar.path,
+                tools_spec=tools_spec
+            )
+            )
+
+    def build(self, spec, prefix):
+        x_py = Executable('./x.py')
```

Comment:
I would probably just do:
```python
self.python('x.py', 'build', ...)
```
That way it uses the same Python in the Spack DAG, regardless of PATH or shebang.

Reply:
Oh, huh, why aren't the environment changes in the context of the package sufficient to pick up the dependent python?

Reply:
Also, this seems to require `PythonPackage`? Is that right?

Reply:
> why aren't the environment changes in the context of the package sufficient to pick up the dependent python?

The shebang of `x.py` looks like:
```bash
#!/usr/bin/env python                                                           
```
However, if you build `python@3:~pythoncmd`, or use an external Python built in the same way, it won't have a `python` executable, only a `python3` executable, so `./x.py` will end up using the system `python` if one exists. There are newer Ubuntu versions that don't come with any `python`, only `python3` and `python2`. Calling the script with the right executable will prevent that.

> Also, this seems to require PythonPackage? Is that right?

It _shouldn't_. `self.python` should be set [here](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/python/package.py#L774) for all packages that depend on Python.

Reply:
Oooh, OK.

Reply:
```
==> Error: Failed to install rust due to ChildError: AttributeError: 'Rust' object has no attribute 'python'
/yellow/usr/projects/eap/users/agaspar/spack-dev/var/spack/repos/builtin/packages/rust/package.py:450, in build:
        449    def build(self, spec, prefix):
  >>    450        self.python('./x.py', 'build', extra_env={
        451            # vendored libgit2 wasn't correctly building (couldn't find the
        452            # vendored libssh2), so let's just have spack build it
        453            'LIBSSH2_SYS_USE_PKG_CONFIG': '1',
```

Reply:
Try `python(...)` instead of `self.python(...)`

Reply:
Let me know if there's a better way to do it, but it doesn't seem like it's working for me...

Reply:
`python(...)` should work, a lot of packages use that.

Reply:
oh, sorry, didn't see your reply.

Reply:
I'm not a python expert, but I'm guessing this is because it's a class member rather than an instance member?

Reply:
That seemed to fix it.